mlir/lib/Tools/PDLL/Parser/Parser.cpp

```diff
@@ -1915,28 +1964,180 @@ FailureOr<ast::Expr *> Parser::parseEqualityExpr() {
 FailureOr<ast::Expr *> Parser::parseRelationExpr() { return parseAddSubExpr(); }
 
 FailureOr<ast::Expr *> Parser::parseAddSubExpr() {
-  auto lhs = parseMulDivExpr();
+  auto lhs = parseMulDivModExpr();
+  if (failed(lhs))
+    return failure();
+
+  for (;;) {
+    switch (curToken.getKind()) {
+    case Token::add: {
+      consumeToken();
+      auto rhs = parseMulDivModExpr();
+      if (failed(rhs))
+        return failure();
+      SmallVector<ast::Expr *> args{*lhs, *rhs};
+
+      // Check if it is in rewrite section but not in the let statement
+      bool inRewriteSection = parserContext == ParserContext::Rewrite;
+      if (inRewriteSection &&
+          nativeOperatorContext == NativeOperatorContext::Generic)
+        return emitError(
+            "nodiscard rule for add operator is applied in rewrite section");
```

Comment:
I think this message could be improved. Maybe something like:
```suggestion
            "cannot evaluate add operator in rewrite section. Assign to a variable with `let`");
```
I am not very fan with my suggestion, so feel free to improve it. 😄  maybe someone else has a better suggestion.